File: leetcode/Python/0718-m-maximum-length-of-repeated-subarray.py
"""
DP
dp[i][j]: the longest common prefix of A[i:] and B[j:]
If A[i] == B[j], dp[i][j] = dp[i + 1][j + 1] + 1
answer: max(dp[i][j])
Time: O(mn)
Space: O(mn)
"""
class Solution:
    def findLength(self, A: List[int], B: List[int]) -> int:
        dp = [[0] * (len(B) + 1) for _ in range(len(A) + 1)]
        for i in range(len(A) - 1, -1, -1):
            for j in range(len(B) - 1, -1, -1):
                if A[i] == B[j]:
                    dp[i][j] = dp[i + 1][j + 1] + 1
        return max(max(row) for row in dp)

File: leetcode/C++/0718-m-maximum-length-of-repeated-subarray.cpp
// DP
// Time: O(mn)
// Space: O(mn)
class Solution {
public:
    int findLength(vector<int>& A, vector<int>& B) {
        int m = A.size(), n = B.size(), res = 0;
        vector<vector<int>> dp(m, vector<int>(n));
        for(int i = 0; i < m; ++i)
            for(int j = 0; j < n; ++j)
                if(A[i] == B[j])
                {
                    dp[i][j] = (i && j ? dp[i - 1][j - 1] : 0) + 1;
                    res = max(res, dp[i][j]);
                }
        return res;
    }
};